### DISCUSSION TITLE
[Support] websocket中如何标记连接用户？

### DISCUSSION THREAD

**Discussion by elicip** [midwayjs/midway#4034] [Q&A]
### Describe the problem（描述问题）

使用的是@midwayjs/ws@3，我在connect的时候使用setAttr给socket设置了一个uuid，但是在别的service中使用this.wsApp.clients如何定位到这个socket呢？

```
 @OnWSConnection()
  async onConnectionMethod(socket: Context, request: http.IncomingMessage) {
    // 给当前连接的客户端设置一个唯一id
    socket.setAttr('uuid', uuid());
  }
```

或者说这个方法是错误的，那么应该如何实践呢？我是否需要额外创建一个单例service去保存每个socket会话啊

### Midway Versions（Midway 版本）

npx midway-version@latest

**Comment by czy88840616** [comment]
请求作用域的 service，注入 ctx 即可。

**Reply by elicip** [reply]
这样还是无法从this.wsApp.clients中筛选出标记的那个客户端吧？this.wsApp.clients.forEach(ws => {ws并没有getAttr方法})

**Reply by czy88840616** [reply]
一个 ctx 就是对应了一个固定的 socket 呀，不用筛选，也不会串的。

**Reply by elicip** [reply]
抱歉，我没理解你的意思。
我作为服务端，需要给其中一个标记了的客户端socket发消息。比如我有两台硬件设备连接上了我的websocket server，我分别对他们做标记比如设备A和设备B，我管理后台写一个service给设备A下发指令，这时候我应该首先去筛选出他对应的socket会话才能给他发消息吧？

**Reply by czy88840616** [reply]
两种场景：
*  1、如果是消息链路，比如设备 A 发送了消息，server 接到然后调用了 Service，那么 service 中 inject 的 ctx 即为这个 socket，直接发送消息就好。
*  2、如果非消息链路（如后台任务），或者是获取非当前链路的 socket，则需要在连接时是将 client 存起来，你可以存到 app.setAttr 中。

**Reply by elicip** [reply]
1. 嗯，关于场景一我明白
2. 看来还是得自己实现将client存储起来，我看到this.wsApp.clients，还以为是可以复用直接在这个里面标记。midwayjs的websocket是不是还得处理多进程下存储client的问题？

**Reply by czy88840616** [reply]
wsApp.clients 是ws 自带的存了所有客户端，但是无法区分，广播的时候还行。至于多进程就是另外的问题，毕竟每个客户端连接只会连接到一个进程上。

**Reply by layyback** [reply]
可以直接使用
    wsApp.clients.forEach((client) => {
      const uuid = client.getAttr('uuid');
      .....
    });
我试了可行的

（多进程需要进程间通信后，再各自遍历所有实例）
